src/lib/logstore/log_dev.cpp

```diff
@@ -399,233 +399,152 @@ bool LogDev::flush_if_needed(int64_t threshold_size) {
+        THIS_LOGDEV_LOG(TRACE, "Log idx {} last_flush_idx {} prepare flush failed", new_idx, m_last_flush_idx);
+        return false;
     }
-#endif
-
-    // if (has_data_service() && data_service().is_fsync_needed()) {
-    //     data_service().fsync([this, lg]() { do_flush_write(lg); })
-    // } else {
-    //     do_flush_write(lg);
-    // }
-    do_flush_write(lg);
-}
+    auto sz = m_pending_flush_size.fetch_sub(lg->actual_data_size(), std::memory_order_relaxed);
+    HS_REL_ASSERT_GE((sz - lg->actual_data_size()), 0, "size {} lg size{}", sz, lg->actual_data_size());
+    off_t offset = m_vdev_jd->alloc_next_append_blk(lg->header()->total_size());
+    lg->m_log_dev_offset = offset;
+
+    HS_REL_ASSERT_NE(lg->m_log_dev_offset, INVALID_OFFSET, "log dev is full");
+    THIS_LOGDEV_LOG(TRACE, "Flushing log group data size={} at offset=0x{} log_group={}", lg->actual_data_size(),
+                    to_hex(offset), *lg);
 
-void LogDev::do_flush_write(LogGroup* lg) {
     HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_flush_records_distribution, lg->nrecords());
     HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_flush_size_distribution, lg->actual_data_size());
-    THIS_LOGDEV_LOG(TRACE, "vdev offset={} log group total size={}", lg->m_log_dev_offset, lg->header()->total_size());
 
     // write log
-    m_vdev_jd->async_pwritev(lg->iovecs().data(), int_cast(lg->iovecs().size()), lg->m_log_dev_offset)
-        .thenValue([this, lg](auto) { on_flush_completion(lg); });
+    m_vdev_jd->sync_pwritev(lg->iovecs().data(), int_cast(lg->iovecs().size()), lg->m_log_dev_offset);
+
+    on_flush_completion(lg);
+    return true;
 }
 
 void LogDev::on_flush_completion(LogGroup* lg) {
-    lg->m_flush_finish_time = Clock::now();
-    lg->m_post_flush_msg_rcvd_time = Clock::now();
+    auto done_time = Clock::now();
     THIS_LOGDEV_LOG(TRACE, "Flush completed for logid[{} - {}]", lg->m_flush_log_idx_from, lg->m_flush_log_idx_upto);
 
     m_log_records->complete(lg->m_flush_log_idx_from, lg->m_flush_log_idx_upto);
     m_last_flush_idx = lg->m_flush_log_idx_upto;
-    const auto flush_ld_key = logdev_key{m_last_flush_idx, lg->m_log_dev_offset + lg->header()->total_size()};
     m_last_crc = lg->header()->cur_grp_crc;
 
     auto from_indx = lg->m_flush_log_idx_from;
     auto upto_indx = lg->m_flush_log_idx_upto;
     auto dev_offset = lg->m_log_dev_offset;
     for (auto idx = from_indx; idx <= upto_indx; ++idx) {
         auto& record = m_log_records->at(idx);
-        on_io_completion(record.store_id, logdev_key{idx, dev_offset}, flush_ld_key, upto_indx - idx, record.context);
+        logstore_req* req = s_cast< logstore_req* >(record.context);
+        HomeLogStore* log_store = req->log_store;
+        HS_LOG_ASSERT_EQ(log_store->get_store_id(), record.store_id,
+                         "Expecting store id in log store and flush completion to match");
+
+        log_store->on_write_completion(req, logdev_key{idx, dev_offset});
```

Comment:
can we do the callback w/o holding flush_guard?

Reply:
done